File: examples/typescript/yds/function/producer.ts
import { Http } from "@yandex-cloud/function-types/dist/src/http";
import { Driver } from "@ydbjs/core";
import { MetadataCredentialsProvider } from "@ydbjs/auth/metadata";
import { createTopicWriter } from "@ydbjs/topic/writer";
import { AccessTokenCredentialsProvider } from "@ydbjs/auth/access-token";

/**
 * Producer function that receives HTTP requests and writes messages to YDS topic.
 *
 * Expected request body:
 * {
 *   "message": "string",
 *   "user_id": "string",
 *   "action": "string"
 * }
 */
export async function handler(event: Http.Event): Promise<Http.Result> {
  try {
    // Parse request body
    let body: string = event.body || "{}";
    if (event.isBase64Encoded) {
      body = Buffer.from(body, "base64").toString("utf-8");
    }

    const data = JSON.parse(body);

    // Validate required fields
    const requiredFields = ["message", "user_id", "action"];
    for (const field of requiredFields) {
      if (!(field in data)) {
        return {
          statusCode: 400,
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ error: `Missing required field: ${field}` }),
        };
      }
    }

    // Get environment variables
    const ydbEndpoint = process.env.YDB_ENDPOINT;
    const ydsTopicPath = process.env.YDS_TOPIC_PATH;

    if (!ydbEndpoint || !ydsTopicPath) {
      console.error(
        "Missing environment variables: YDB_ENDPOINT or YDS_TOPIC_PATH"
      );
      return {
        statusCode: 500,
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ error: "Configuration error" }),
      };
    }

    // Prepare message with timestamp
    const message = {
      message: data.message,
      user_id: data.user_id,
      action: data.action,
      timestamp: new Date().toISOString(),
    };

    // Connect to YDB and write to topic
    console.log(`Connecting to YDB: ${ydbEndpoint}`);
    console.log(`Topic path: ${ydsTopicPath}`);

    const iamToken = process.env.YDB_IAM_TOKEN;
    let credentialsProvider;
    if (iamToken) {
      console.log("Using IAM token for authentication");
      credentialsProvider = new AccessTokenCredentialsProvider({
        token: iamToken,
      });
    } else {
      console.log("No IAM token provided; using default credentials provider");
      credentialsProvider = new MetadataCredentialsProvider();
    }

    const driver = new Driver(ydbEndpoint, {
      credentialsProvider,
    });

    try {
      console.log("Waiting for driver to be ready...");
      await driver.ready();
      console.log("Driver is ready");

      // Create topic writer using the documented API
      console.log("Creating topic writer...");
      await using writer = createTopicWriter(driver, {
        topic: ydsTopicPath,
        producer: "producer-ts",
      });
      console.log("Topic writer created");

      // Write message to topic
      const messageData = JSON.stringify(message);
      console.log(`Writing message: ${messageData}`);
      const seqNo = writer.write(
        new TextEncoder().encode(messageData)
      );
      console.log(`Message queued with sequence number: ${seqNo}`);

      console.log("Flushing writer...");
      await writer.flush();
      console.log("Writer flushed successfully");

      console.log(`Message sent to YDS topic: ${JSON.stringify(message)}`);

      return {
        statusCode: 200,
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          status: "success",
          message: "Message sent to YDS topic",
          data: message,
          sequenceNumber: seqNo.toString(),
        }),
      };
    } catch (writeError) {
      console.error("Error during YDB write operation:", writeError);
      console.error(
        "Error stack:",
        writeError instanceof Error ? writeError.stack : String(writeError)
      );
      throw writeError;
    } finally {
      // Close driver
      console.log("Closing driver...");
      await driver.close();
      console.log("Driver closed");
    }
  } catch (error) {
    if (error instanceof SyntaxError) {
      console.error("Invalid JSON in request body:", error);
      return {
        statusCode: 400,
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ error: "Invalid JSON format" }),
      };
    }

    console.error("Error processing request:", error);
    return {
      statusCode: 500,
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        error: error instanceof Error ? error.message : String(error),
      }),
    };
  }
}

File: examples/python/yds/function/producer.py
import json
import os
import logging
from datetime import datetime
import ydb
import ydb.iam

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)


def producer_handler(event, context):
    """
    Producer function that receives HTTP requests and writes messages to YDS topic.

    Expected request body:
    {
        "message": "string",
        "user_id": "string",
        "action": "string"
    }
    """
    try:
        # Parse request body
        body = event.get('body', '{}')
        if event.get('isBase64Encoded', False):
            import base64
            body = base64.b64decode(body).decode('utf-8')

        data = json.loads(body)

        # Validate required fields
        required_fields = ['message', 'user_id', 'action']
        for field in required_fields:
            if field not in data:
                return {
                    'statusCode': 400,
                    'body': json.dumps({'error': f'Missing required field: {field}'})
                }

        # Get environment variables
        ydb_endpoint = os.environ.get('YDB_ENDPOINT')
        yds_topic_path = os.environ.get('YDS_TOPIC_PATH')

        if not ydb_endpoint or not yds_topic_path:
            logger.error("Missing environment variables: YDB_ENDPOINT or YDS_TOPIC_PATH")
            return {
                'statusCode': 500,
                'body': json.dumps({'error': 'Configuration error'})
            }

        # Prepare message with timestamp
        message = {
            'message': data['message'],
            'user_id': data['user_id'],
            'action': data['action'],
            'timestamp': datetime.utcnow().isoformat()
        }

        # Parse YDB endpoint to extract database path
        # Format: grpcs://host:port/?database=/path/to/db
        database_path = yds_topic_path.rsplit('/', 1)[0]  # Remove topic name to get database path

        # Connect to YDB and write to topic
        driver_config = ydb.DriverConfig(
            ydb_endpoint,
            database=database_path,
            credentials=ydb.iam.MetadataUrlCredentials(),
        )

        with ydb.Driver(driver_config) as driver:
            driver.wait(timeout=5, fail_fast=True)

            # Write message to topic
            topic_path = yds_topic_path.split(database_path)[-1].lstrip('/')  # Get relative topic path
            with driver.topic_client.writer(topic_path) as writer:
                writer.write(
                    ydb.TopicWriterMessage(
                        data=json.dumps(message).encode('utf-8')
                    )
                )
                writer.flush()

        logger.info(f"Message sent to YDS topic: {message}")

        return {
            'statusCode': 200,
            'body': json.dumps({
                'status': 'success',
                'message': 'Message sent to YDS topic',
                'data': message
            })
        }

    except json.JSONDecodeError as e:
        logger.error(f"Invalid JSON in request body: {e}")
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Invalid JSON format'})
        }
    except Exception as e:
        logger.error(f"Error processing request: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }
